pkg/endpoint/policy.go

```diff
@@ -516,19 +516,21 @@ func (e *Endpoint) regeneratePolicy(owner Owner, opts models.ConfigurationMap) (
 		e.syncPolicyMapController()
 	}
 
-	// If no policy or options change occurred for this endpoint then the endpoint is
-	// already running the latest revision, otherwise we have to wait for
-	// the regeneration of the endpoint to complete.
-	if !policyChanged && !optsChanged {
-		e.setPolicyRevision(revision)
-	}
-
 	e.getLogger().WithFields(logrus.Fields{
 		"policyChanged":       policyChanged,
 		"optsChanged":         optsChanged,
 		"policyRevision.next": e.nextPolicyRevision,
 	}).Debug("Done regenerating")
 
+	// If no policy or options change occurred for this endpoint then the
+	// endpoint is already running the latest revision; otherwise we have to
+	// wait for the regeneration of the endpoint to complete. This allows us
+	// to signal that regeneration of BPF is not necessary.
+	if !policyChanged && !optsChanged {
+		e.setPolicyRevision(revision)
+		return false, nil
+	}
+
 	needToRegenerateBPF := optsChanged || policyChanged || e.nextPolicyRevision > e.policyRevision
```

Comment:
You are changing behavior here. Are you sure it is safe? Do we no longer need to regenerate if nextPolicyRevision > policyRevision?

Reply:
On L513, we set `e.NextPolicyRevision = revision`. Since we are calling `e.setPolicyRevision(revision)`, that means that we are guaranteed to have `e.nextPolicyRevision == e.policyRevision`. Given that the change in logic is `if !policychanged && !optsChanged`, which means that `policyChanged` and `optsChanged` are both false, then `needToRegenerateBPF := optsChanged || policyChanged || e.nextPolicyRevision > e.policyRevision` equates to `false || false || false`, which will result in the function returning `false, nil`. I did this because I felt like it made it more clear. But, i can see why this is confusing, and we after seeing this, agree that we should have return logic for whether we need to regenerate BPF stored explicitly in the `needToRegenerateBPF` variable, as it will make the code more readable. Will fix in next upload.

Reply:
Fixed.